features/api-layer/src/main/java/org/opennms/features/apilayer/collectors/ServiceCollectorImpl.java

```diff
@@ -0,0 +1,183 @@
+        ServiceCollector serviceCollector = serviceCollectorFactory.createCollector();
+        serviceCollector.initialize();
+        CollectionRequest collectionRequest = new CollectionRequestImpl(agent);
+        CompletableFuture<CollectionSet> future = serviceCollector.collect(collectionRequest, parameters);
+        CollectionSetBuilder builder = new CollectionSetBuilder(agent);
+        try {
+            CollectionSet collectionSet = future.get();
+            if (collectionSet.getStatus().equals(CollectionSet.Status.FAILED)) {
+                return  builder.withTimestamp(new Date(collectionSet.getTimeStamp()))
+                        .withStatus(CollectionStatus.FAILED).build();
+            }
+            for (CollectionSetResource collectionSetResource : collectionSet.getCollectionSetResources()) {
+                Resource resource = collectionSetResource.getResource();
+                if (resource == null) {
+                    continue;
+                }
+                if (resource.getResourceType().equals(Resource.Type.NODE)) {
+                    NodeResource nodeResource = (NodeResource) resource;
+                    NodeLevelResource nodeLevelResource = new NodeLevelResource(nodeResource.getNodeId());
+                    addAttributes(collectionSetResource, builder, nodeLevelResource);
+                } else if (resource.getResourceType().equals(Resource.Type.INTERFACE)) {
+                    IpInterfaceResource ipResource = (IpInterfaceResource) resource;
+                    NodeLevelResource nodeLevelResource = new NodeLevelResource(ipResource.getNodeResource().getNodeId());
+                    InterfaceLevelResource interfaceLevelResource = new InterfaceLevelResource(nodeLevelResource, ipResource.getInstance());
+                    addAttributes(collectionSetResource, builder, interfaceLevelResource);
+                } else if (resource.getResourceType().equals(Resource.Type.GENERIC)) {
+                    GenericTypeResource genericTypeResource = (GenericTypeResource) resource;
+                    NodeLevelResource nodeLevelResource = new NodeLevelResource(genericTypeResource.getNodeResource().getNodeId());
+                    // TODO: Get ResourceType from ResourceTypesDao
+                    org.opennms.netmgt.collection.support.builder.GenericTypeResource genericResource =
+                            new org.opennms.netmgt.collection.support.builder.GenericTypeResource(nodeLevelResource, null, genericTypeResource.getInstance());
+                    addAttributes(collectionSetResource, builder, genericResource);
+                }
+            }
+            builder.withTimestamp(new Date(collectionSet.getTimeStamp()));
+            return builder.build();
+
+        } catch (InterruptedException | ExecutionException e) {
+            LOG.error("collect failed with ", e);
```

Comment:
We should probably rethrow this as a CollectionException instead of just returning `CollectionStatus.UNKNOWN`.